src/combinatorics/ssyt.lean

```diff
@@ -0,0 +1,103 @@
+/-
+Copyright (c) 2022 Jake Levinson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jake Levinson
+-/
+import combinatorics.young_diagram
+
+/-!
+# Semistandard Young tableaux
+
+A semistandard Young tableau is a filling of a Young diagram by natural numbers, such that
+the entries are weakly increasing left-to-right along rows (i.e. for fixed `i`), and
+strictly-increasing top-to-bottom along columns (i.e. for fixed `j`).
+
+An example of an SSYT of shape `μ = [4, 2, 1]` is:
+
+```text
+0 0 0 2
+1 1
+2
+```
+
+We represent an SSYT as a function `ℕ → ℕ → ℕ`, which is required to be zero for all pairs
+`(i, j) ∉ μ` and to satisfy the row-weak and column-strict conditions on `μ`.
+
+
+## Main definitions
+
+- `ssyt (μ : young_diagram)` : semistandard Young tableaux of shape `μ`. There is
+  a `has_coe_to_fun` instance such that `T i j` is value of the `(i, j)` entry of the SSYT `T`.
+- `ssyt.highest_weight (μ : young_diagram)`: the semistandard Young tableau whose `i`th row
+  consists entirely of `i`s, for each `i`.
+
+## Tags
+
+Semistandard Young tableau
+
+## References
+
+<https://en.wikipedia.org/wiki/Young_tableau>
+
+-/
+
+/-- A semistandard Young tableau (SSYT) is a filling of the cells of a Young diagram by natural
+numbers, such that the entries in each row are weakly increasing (left to right), and the entries
+in each column are strictly increasing (top to bottom).
+
+Here, an SSYT is represented as an unrestricted function `ℕ → ℕ → ℕ` that, for reasons of extensionality,
+is required to vanish outside `μ`. --/
+structure ssyt (μ : young_diagram) :=
+(entry : ℕ → ℕ → ℕ)
+(row_weak : ∀ {i j1 j2 : ℕ}, j1 < j2 → (i, j2) ∈ μ → entry i j1 ≤ entry i j2)
+(col_strict : ∀ {i1 i2 j : ℕ}, i1 < i2 → (i2, j) ∈ μ → entry i1 j < entry i2 j)
+(zeros' : ∀ {i j}, (i, j) ∉ μ → entry i j = 0)
+
+namespace ssyt
+
+instance fun_like {μ : young_diagram} : fun_like (ssyt μ) ℕ (λ _, ℕ → ℕ) :=
+{ coe := ssyt.entry,
+  coe_injective' := λ T T' h, by { cases T, cases T', congr' } }
+
+/-- Helper instance for when there's too many metavariables to apply
+`fun_like.has_coe_to_fun` directly. -/
+instance {μ : young_diagram} : has_coe_to_fun (ssyt μ) (λ _, ℕ → ℕ → ℕ) :=
+fun_like.has_coe_to_fun
+
+@[simp] lemma to_fun_eq_coe {μ : young_diagram} {T : ssyt μ} : T.entry = (T : ℕ → ℕ → ℕ) := rfl
+
+@[ext] theorem ext {μ : young_diagram} {T T' : ssyt μ} (h : ∀ i j, T i j = T' i j) : T = T' :=
+  fun_like.ext T T' (λ x, by { funext, apply h })
+
+/-- Copy of an `ssyt μ` with a new `entry` equal to the old one. Useful to fix definitional
+equalities. -/
+protected def copy {μ : young_diagram} {T : ssyt μ} (entry' : ℕ → ℕ → ℕ) (h : entry' = T) :
+ssyt μ :=
+{ entry := entry',
+  row_weak := h.symm ▸ T.row_weak,
+  col_strict := h.symm ▸ T.col_strict,
+  zeros' := h.symm ▸ T.zeros' }
```

Comment:
A thought for a future PR:

Are there cases where you want to rewrite `μ`? If so, you probably want to have an additional `protected def cast {μ μ' : young_diagram} (T : ssyt μ) (h : μ = μ') : ssyt μ' := eq.rec h T` (I didn't check this in Lean, so the `eq.rec` term might be wrong).

Reply:
Actually, I think the purpose of this copy pattern is basically to avoid ever rewriting the `μ` term in an `ssyt μ`. I remember this causing me some headaches with the recursive definition of RSK insertion.

You were actually the one to suggest the copy pattern to me in [this thread](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/dependent.20structures). I think it worked pretty well generally.